### PR TITLE
Airport

### PR DESCRIPTION
**Pull request by wvarty**
### What is Airport?
Airport provides a new firmware option that allows you to turn a regular ExpressLRS TX and RX pair into a bi-directional transparent serial link, over the air.
The TX module is intended to connect via USB to a laptop, and the RX connects to a free UART on your FC, as per usual. This then allows serial comms between the peers, in any protocol you wish to use, for example: MAVLINK (Ardupilot), MSP (Betaflight and INAV), or any other telemetry protocol you might use.

### What hardware do I need to try this?
The Airport option completely replaces the RC link, and repurposes it as a data link. This means you need to run 2x TXs and 2x RXs on the ground and air, respectively. One TX+RX pair sends your normal RC link data, just as it does it does now, and the other TX+RX pair sends the serial data.
airport-diagram

The feature should compile for any supported ELRS target, however the TX module makes sense to be an external one, so you can connect it via USB to a laptop, and it is helpful to have an OLED screen on the TX module so you can change packet rates / power output without plugging it into a handset JR bay. It should compile for 868/900/2.4 targets, on both STM and ESP architectures, so all RXs with a full UART (both tx and rx pads) should work.

### What baud rate should I use?
The baud rate you select to use should be the same on the TX and RX, and it should also be slow enough such that the packet rate can send the data.
For example, at a packet rate of 500hz LoRa, and a 1:2 hard-coded TLM rate, we can send 500hz / 2 * 5bytes = 1250bytes/s.
A baud rate of 19200 allows for a data rate of 19200 * 0.8 / 8bits/byte = 1920bytes/s. This is too much data for 500hz, since 1920 > 1250.
A baud rate of 9600 allows for a data rate of 9600* 0.8 / 8bits/byte = 960bytes/s. This is an acceptable rate since 960 < 1250.

### How do I test this PR?
Long term, this will be an option you can enable in the ELRS Configurator.
For now though, the user define should be used via the manual FW build process (read: dev friendly only ATM).
Enable the `-DUSE_AIRPORT_AT_BAUD=9600` option by removing the # at the beginning of the line in `user_defines.txt`, and set your binding phrase user define to something different to what you use for normal ELRS RC control.
Build and flash a TX and RX pair using the above options, and connect RX to a spare UART, and TX to a PC via USB.
Configure your FC to talk via whatever protocol you prefer, for example if using Ardupilot, configure the UART for MAVLINK protocol.
Open the connection from the Configurator / GCS system on the PC.
Profit!

### CODE REVIEW

**Comment by wvarty** [comment]
Few minor things yet to fix here, but this is starting to look pretty good:

1. RF mode changes aren't working correctly if the link is already connected. Changing packet rate on the TX currently requires a power cycle of the RX to regain link.
2. Seems to be an issue with LUA not showing the "Other devices" menu option when a RX is connected.
3. Connecting to TX to both a USB cable and a handset then, disconnecting the CRSF connection only (while keeping USB attached and providing power) causes the link to drop (suspect this is a trigger due to the loss of CRSF connection)

**Comment by YI-BOYANG** [comment]
This feature requires two ELRS TX to implement, so why not just buy a SIYI FM30. (⊙ˍ⊙)

**Comment by YI-BOYANG** [comment]
This feature requires two ELRS TX to implement, so why not just buy a SIYI FM30. (⊙ˍ⊙)

**Comment by wvarty** [comment]
> This feature requires two ELRS TX to implement, so why not just buy a SIYI FM30. (⊙ˍ⊙)

Yep, there's a target for flashing ELRS to the FM30, so that should work as the second TX module, although I have mainly been testing on the ESP32 based TX targets, so YMMV.

**Comment by CREASE-gum-EAR** [comment]
Can this mode also run CLI commands and firmware updates on the FC? I'm currently using bluetooth for this. The ELRS Wifi can't do the CLI commands and firmware updates.

**Comment by Dropsleutel** [comment]
Is there any way of synchronization in hopping schemes, between the onboard Rx and the Airport Tx ?  Very strong signals on adjescent channels may affect the range ? So the airport not reduces the Rx LQ at longer distances ?

**Comment by wvarty** [comment]
> Is there any way of synchronization in hopping schemes, between the onboard Rx and the Airport Tx ? Very strong signals on adjescent channels may affect the range ? So the airport not reduces the Rx LQ at longer distances ?

Unfortunately no. The second link is basically the same a having someone else fly standing next to you, which relies on FHSS to avoid collisions.

**Comment by wvarty** [comment]
Taking this out of draft, as I think it is ready for review and testing.

**Comment by YohanHadji** [comment]
Trying to test this feature with a BETAFPV 868 nano RX and micro TX. Both flashed with `airport-rebase` using the configurator. The RX over wifi, the TX over UART. Sending packets ("hello world") with a Teensy 4.1 at 9600 baud (through a logic level shifter, teensy has a 3.3V level and I suspect the nano rx to have a 5v logic level). Bind phrase set correctly on both sides, `-DUSE_AIRPORT_AT_BAUD=9600` set too. Packet rate set to 200Hz using the interface of the micro TX. 

Connection between the RX and TX is good (solid led on RX and "connected" text on the TX). Reading the serial port of the micro TX on a PC via the USB interface of the micro TX module, bytes are coming in at the good frequency, but empty of any data (all bytes are "null" ascii code). I confirmed with a second teensy that the first teensy is sending the code over UART correctly. 

What would you recommend to try next?

**Comment by YohanHadji** [comment]
Fixed, the `-DUSE_AIRPORT_AT_BAUD=9600` setting is not the only baud setting to set correctly, this one on the screenshot too needs to se set (on the RX) to the same baud for it to work. I think the two settings are in conflict somewhere? I'll do more testing. 

Capture d’écran 2022-11-30 à 20 32 34

**Comment by CorruptName** [comment]
This isn't working for me.
Using HappyModel EP1 RX I get:

`TARGET MISMATCH
Current target: HAPPYMODEL_EP_2400_RX.

It looks like you are flashing firmware with a different name to the current  firmware.  This sometimes happens because the hardware was flashed from the factory with an early version that has a different name. Or it may have even changed between major releases.

Please double check you are uploading the correct target, then proceed with 'Flash Anyway'.
WIFI upload FAILED!
*** [upload] Error -2`


If I try to "Force Flash" I get "Not enough space" error.
Am I doing something wrong?


[ExpressLRSBuildLog_20230114082526133.txt](https://github.com/ExpressLRS/ExpressLRS/files/10416987/ExpressLRSBuildLog_20230114082526133.txt)

**Comment by CorruptName** [comment]
I tried with Axis Thor 2.4 RX and it flashed fine so I don't thing it's my build environment.

**Comment by vrquaeler** [comment]
Hi, I just discovered this PR. Started to experiment with.  Got it working on a Namimno Flash 2.4G OLED TX and a 2.4G Vantac Lite module. Bought a Jumper NANO TX module. Seemed the perfect form factor for my (ArduPlane related) application, it's small and with a display and a 5 way button. Can't get it work. After flashing "Airport-rebase" it does nothing. No visible RB-led, nothing on the display and no wifi. I am still able to flash the module with regular V2.5.1 or on V3.1.1 or V3.1.2. Any ideas or got I bitten by a bug?

**Comment by TimoLehnertz** [comment]
Is there any way to have multiple receivers send telemetry to one transmitter?

**Comment by teslahed** [comment]
I'm trying to compile airport to give it a go. I am sure I managed it about 2 weeks ago but I can't get it to work anymore despite there seemingly being no change to my configuration since then.

I get this error message;
Error: Invalid '/home/chris/Code/ExpressLRS-airport-rebase/src/platformio.ini' (project configuration file): 'No option 'bf_upload_command' in section: 'env_common_esp82xx''

and a very similar one if I use visual studio code instead of the expressLRS app to compile it. 

Anyone got any ideas or suggestions on this one? Thanks.

**Comment by TimoLehnertz** [comment]
I am having the exact same issue. Same in vscode and the configurator. Only with the airport rebase branch, tho. The airport unified branch builds fine

**Comment by teslahed** [comment]
> I am having the exact same issue. Same in vscode and the configurator. Only with the airport rebase branch, tho. The airport unified branch builds fine

This might sound stupid but what's the difference between the airport rebase branch and the airport unified branch? Maybe I should be using the later?

**Comment by CorruptName** [comment]
> I'm trying to compile airport to give it a go. I am sure I managed it about 2 weeks ago but I can't get it to work anymore despite there seemingly being no change to my configuration since then.
> 
> I get this error message; Error: Invalid '/home/chris/Code/ExpressLRS-airport-rebase/src/platformio.ini' (project configuration file): 'No option 'bf_upload_command' in section: 'env_common_esp82xx''
> 
> and a very similar one if I use visual studio code instead of the expressLRS app to compile it.
> 
> Anyone got any ideas or suggestions on this one? Thanks.

Same Problem.

**Comment by MingchenZhang** [comment]
> I'm trying to compile airport to give it a go. I am sure I managed it about 2 weeks ago but I can't get it to work anymore despite there seemingly being no change to my configuration since then.
> 
> I get this error message; Error: Invalid '/home/chris/Code/ExpressLRS-airport-rebase/src/platformio.ini' (project configuration file): 'No option 'bf_upload_command' in section: 'env_common_esp82xx''
> 
> and a very similar one if I use visual studio code instead of the expressLRS app to compile it.
> 
> Anyone got any ideas or suggestions on this one? Thanks.

I guess the platformIO/configurator updated some of its behavior, and was fixed in the later revision. I rebased the airport-rebase branch onto the 3.x.x-maintenance branch, and it compiles in configurator again.

**Comment by MingchenZhang** [comment]
> Fixed, the `-DUSE_AIRPORT_AT_BAUD=9600` setting is not the only baud setting to set correctly, this one on the screenshot too needs to se set (on the RX) to the same baud for it to work. I think the two settings are in conflict somewhere? I'll do more testing.
> 
> EDIT: the baud setting in the wifi portal seems to be completely overwriting the baud set with the `-DUSE_AIRPORT_AT_BAUD=9600` setting.
> 
> Capture d’écran 2022-11-30 à 20 32 34

Nice observation. That is the problem made it behave weirdly in my test. 
https://github.com/ExpressLRS/ExpressLRS/blob/1731408a17890e9cc4ecc6f231582e7e26761971/src/lib/OPTIONS/options.cpp#L314
This line overwrite the baudrate. Not sure if it is the best to always override baudrate stored in the flash silently by a compilation flag, but changing it to the USE_AIRPORT_AT_BAUD there will make USE_AIRPORT_AT_BAUD function as intended.

**Comment by teslahed** [comment]
> > I'm trying to compile airport to give it a go. I am sure I managed it about 2 weeks ago but I can't get it to work anymore despite there seemingly being no change to my configuration since then.
> > I get this error message; Error: Invalid '/home/chris/Code/ExpressLRS-airport-rebase/src/platformio.ini' (project configuration file): 'No option 'bf_upload_command' in section: 'env_common_esp82xx''
> > and a very similar one if I use visual studio code instead of the expressLRS app to compile it.
> > Anyone got any ideas or suggestions on this one? Thanks.
> 
> I guess the platformIO/configurator updated some of its behavior, and was fixed in the later revision. I rebased the airport-rebase branch onto the 3.x.x-maintenance branch, and it compiles in configurator again.

I've just redownloaded the airport rebase code and nothing seems to have changed. Am I missing something here? I'd really like to get this compiling so i can try using it but I can't at the moment, I'm still getting the exact same error. Which version do you recommend I try downloading and compiling? Thanks.

**Comment by teslahed** [comment]
I think I managed to try and rebase the rebase version of airport back into the 3.x.x-maintenance version of expressLRS as suggested but I got a whole load of conflicts I don't know how to resolve and it failed. Any pointers here would be greatly appreciated. 
CONFLICT (content): Merge conflict in src/lib/FIFO_GENERIC/FIFO_GENERIC.h
error: Failed to merge in the changes.

**Comment by MingchenZhang** [comment]
@teslahed Most of the conflicts don't matter. I have a copy of my rebase in https://github.com/MingchenZhang/ExpressLRS/tree/airport-rebase that could be handy.

**Comment by teslahed** [comment]
> @teslahed Most of the conflicts don't matter. I have a copy of my rebase in https://github.com/MingchenZhang/ExpressLRS/tree/airport-rebase that could be handy.

Thanks for that. I've managed to get your version compiled and installed to my betafpv 868mhz micro TX module and receiver. But I haven't got telemetry working between the two yet. Just a quick question - am I right in thinking I need to use 4800 baud for everything at this lower frequency instead of 9600?

**Comment by teslahed** [comment]
I managed to get acceptable performance out of the airport system on 868mhz at 200hz and 4800 baud. I achieved this by turning things down or off in mavlink that I didn't need. This website helped;
https://ardupilot.org/dev/docs/mavlink-requesting-data.html#using-srx-parameters
I set all my SR1_ parameter values to 0 except for SR_Extra1 = 2, SR_Extra2 = 1 and SR_Position = 1
So now I've got an artificial horizon that updates at about 2 hertz, which is fine for long distance flying, and position data for GPS tracking that also works. 
It's important to set the telemetry rates in Mission Planner to -1 so they don't over ride the SRx settings. 
It takes about 2 to 3 mins to update all my parameters when I first connect via mission planner.

**Comment by teslahed** [comment]
I've recorded a short demo video showing my results so far that I thought I'd post up just in case it helps anyone;
https://www.youtube.com/watch?v=_4WNKI4Q528

**Comment by pkendall64** [comment]
> > Fixed, the `-DUSE_AIRPORT_AT_BAUD=9600` setting is not the only baud setting to set correctly, this one on the screenshot too needs to se set (on the RX) to the same baud for it to work. I think the two settings are in conflict somewhere? I'll do more testing.
> > EDIT: the baud setting in the wifi portal seems to be completely overwriting the baud set with the `-DUSE_AIRPORT_AT_BAUD=9600` setting.

If you have set your bind-phrase via the web-ui or changed any of the other settings then it overrides any flashed values, even when re-flashing new firmware. So it is expected that either you do a full erase before flashing completely different firmware (which airport effectively is) or you have to update that baud-rate in the web-ui.

**Comment by MingchenZhang** [comment]
I went out and give it a try yesterday. Two sets of 2.4 Ghz links, one airport, one regular elrs. Sadly the interference between radiolinks are too much. Both datalink and rc links suffer from weakened signal quality as close as 1km. It could be they are too close and they can hear anything as they a saturated. The distance between two modules on the air side is ~10cm. I will try if increase separation would help. 
It should be much less of an issue for 900mhz and a 2.4ghz links, but i dont have any 900mhz elrs equipments.

**Comment by wvarty** [comment]
Hey all.
Sorry for being a bit AFK on this one.

From the comments above, it looks like there are a few issues people were running into:

1. The baud rate set in the user_define.txt was not overwriting the baud stored in flash from the previous firmware version. This probably needs to be fixed, and @pkendall64 has some ideas on how to address this (we need to ignore the baud in flash if the FW is airport basically).
2. Builds failing - just needed a rebase - thanks @JyeSmith 

In terms of testing, I very much doubt you will see good results using the same frequency on airport as you are on the RC side. 2.4 RC + 2.4 airport isn't going to result in a good time.

I have done some very early testing using 900mhz RC and 2.4G airport, and the results also weren't great (or at least not as good as I'd hoped). This needs more testing.

The final thing I'd like to try is using one airport link for both mavlink AND the RC, by connecting my handset to the GCS PC via USB, and turning on "joystick" in ardu to send RC_OVERRIDE messages from the GCS to the FC. This would mean no crosstalk between RC and TLM links.

**Comment by MingchenZhang** [comment]
Bench test of the using the mavlink to send RC channel is fine for cruise fix wings. The update latency on airport is lower than doing the same on RFD900x. Though RFD900x generally have higher bandwidth. 
Personally I feel RFD900x's range is also not good (at ~24dbm, uplink command not received at ~3km), when last time i used it with 2.4ghz elrs.

File: src/python/o.py [hunk 1]
```diff
@@ -0,0 +1,64 @@
+import sys
```
[the remark this answers is not in the document]
Comment (pkendall64) [reply src/python/o.py:1]: Yeah, that was me inadvertently committing a test program...

**Review by JyeSmith — APPROVED** [review]

**Comment by pkendall64** [comment]
I removed `o.py` (inadvertently committed test code) and also forced the baud-rate on receivers to use the compile time baud rate.

**Review by pkendall64 — APPROVED** [review]
Later on I'd like to revisit this so we can have 3 firmwares; TX, RX & AIRPORT

**Comment by rangler1558** [comment]
Excuse my ignorance as im new to ELRS but is this as simple as doing a build and flash using "airport_rebase" in the configurator on both TX & RX ?

**Comment by kielerrr** [comment]
I'm trying to use MSP for control but the link breaks when trying to read the COM port.
Setup:
- Airport RX(2.4g) on UART2 of the FC 
- 'Configuration/MSP' switch turned on in the Ports tab in Betaflight for UART2. 
- Airport TX is a Radiomaster Ranger connected via USB to the PC as COM5 (Win11)

Using 'Serial Monitor' in ExpressLRS Configurator, when opening COM5 at 9600, the rx loses the solid connections and goes into slow blink, waiting for connection. The tx then freezes and the buttons and oled become unresponsive. Unplugging the USB on the tx and plugging it back in works and the rx reconnects. If I try to read the COM port again, the process repeats.

Using BF Configurator, connecting to COM5 at 9600 does not disconnect the link or freeze the tx but gives this error
```
2023-02-25 @19:51:33 -- Serial port successfully opened with ID: 20
2023-02-25 @19:51:43 -- No configuration received within 10 seconds, communication failed
2023-02-25 @19:51:43 -- Serial port successfully closed
```

Has anyone got airport running with MSP using Win11 and Betaflight?

**Comment by kielerrr** [comment]
> Has anyone got airport running with MSP using Win11 and Betaflight?

I solved this by changing 144000 baud rate to 9600 in the wifi configurator. This is related to [this ](https://github.com/ExpressLRS/ExpressLRS/pull/1904#issuecomment-1332648094)issue

**Comment by boris-gu** [comment]
Do I understand correctly that ELRS RC Link and Airport must operate on different frequencies? Or can I use e.g. 868 for both links?

**Comment by MingchenZhang** [comment]
I tried limiting the airport pair and rc link elrs pair to use different side of the 2.4ghz band. The result is better, the range at which rc link telemetry becomes spotty is further, but there is still a significantly measurable perf drop on the rc link. which does not happen if RFD900X is used with 2.4ghz elrs.

**Comment by wvarty** [comment]
> Do I understand correctly that ELRS RC Link and Airport must operate on different frequencies? Or can I use e.g. 868 for both links?

Not strictly required to be on different frequencies to work, but strongly advised if you want even marginally acceptable performance past about 1km of range.

For comparison, with RC on 2.4G and Airport on 900mhz, 30km was no problem.

**Comment by kf4mot** [comment]
I've got a spare R9M lite tx module. So no usb interface to work with. Is there a uart I could hack into with a usb/serial converter to get this to work? Or another similar mod?

**Comment by teslahed** [comment]
Is it possible to use a tx module as an rx (in the air) to get the same higher power levels in the air as on the ground? A lot of the receivers are quite low on transmit power for the telemetry side of things. I'm not saying I need 1 watt but 100mW or a bit more would be nice. Some of the receivers max out much lower than that I believe.

**Comment by wvarty** [comment]
> If I try to build latest stable, Aiport is not there
AirPort is definitely in the latest `master` branch. Not sure where you are looking? It's not in a release build if that is where you are looking for it? (Currently unreleased)

> I tried to build TX code for RX and so far it fails to boot or work
I don't have any experience with using an RX as a TX, so I can't assist on this question sorry. That isn't really a configuration that is used much by general users, so you are out on your own a bit there in terms of getting it working.

> Would this ever work between two receivers?
In theory, yeah it should, with one flashed as a TX, although that sounds like a very difficult way to try and use the AirPort functionality, as you will have no (easy) way to change packet rate, power, etc. I have also never tried compiling or running it this way, so YMMV.

> How am I suppose to setup Tx if it cannot (easily) connected to handset with LUA script?
You cannot. You have made life extra-hard for yourself using an RX as a TX for this.

> If I use normal Tx module, am I supposed to plug it in radio, setup and take it out?
Yep. This is how I personally use and configure AirPort, and it's even easier if you have a TX module with a screen and 5way.

> If I have Tx module without display, how do I know link is established and running (apart from seeing flowing data)?
The LED on the RX will be solid when you have a link (not flashing)

> If link is established, WiFi is supposed to turn off?
I'm not sure I understand this question? Wifi triggers after the user-defined no-link timeout, same as non-AirPort firmware, and once its in wifi, you need to power cycle to attempt a link.

> Will there be some documentation?
Yep. When AirPort gets released in v3.3, there will be a page on the website to document it.

> Is there any other firmware that would create a transparent serial bridge over SX128x and would compile for ESP8285?
No idea sorry. My simple attempt at it here is the only example I know of that runs on the LoRa hardware.

**Comment by SkylarTheGrey** [comment]
testing this out by compiling in Git Branch and master method, manual defines as follows. R9m and R9mm hardware, Im seeing them connect, but then no serial communication when I have both connected to FTDI adapters. 

Is there supposed to be any sort of activity on traffic indicators or such to tell if one end of the link even sees the serial comms or is the only indication if the data arrives on the other end or not? 

As this is on 915MHz hardware maybe its just never gonna work, but I did try to get settings in the workable range with 200Hz 1:2 set via LUA and compiled with 4800 baud as below (limit should be 4991-5000 based on maths and the LUA). Did I stuff the setup?

-DRegulatory_Domain_FCC_915
-DMY_BINDING_PHRASE="****"
-DFAST_SYNC
-DUSE_AIRPORT_AT_BAUD=4800

**Comment by kieler-zz** [comment]
Nope, you won't see any serial data when airport is working. But when it is working you'll see TX and RX connected and what you write to serial on the TX should pop out as serial data on the RX.

**Comment by kielerrr** [comment]
Nope, you won't see any serial data when airport is working. But when it is working you'll see TX and RX connected and what you write to serial on the TX should pop out as serial data on the RX.

**Comment by SkylarTheGrey** [comment]
you wont see serial when airport is working on what ports? I would expect to see data on the serial ports as that is the goal of the serial bridge. I now strongly suspect an issue with config or compiling as no matter what I probe on the TX I do not see any data that I am pushing to the RX so it is not a pinout issue. Either the RX is seeing the data and it cannot transmit across the link or it is not seeing the data to transmit, either being likely a baud issue?

**Comment by kieler-zz** [comment]
I mean you won't see any serial data if the lines are sitting idle. With non-airport, serial channel data is constantly being exchanged even when not touching sticks and idle. 
Yep, everything has to be the same baud and nothing is at 420000. If you scroll up in this thread a couple people including myself had an issue where 420k baud was being set even with airport enabled. 

In user_defines.txt uncomment the line that sets baud rate and set it to 9600 or whatever you're using. Then reflash TX/RX. Then enable wifi on both devices and check the options tab to make sure baud is set correct. On the TX you might even see a checkbox saying 'use as airport device' and under it you can set baud. 

If you can find me on the elrs discord channel, my username is the same as my github. Send me a message there and I'll see if I can help from there

**Comment by kielerrr** [comment]
I mean you won't see any serial data if the lines are sitting idle. With non-airport, serial channel data is constantly being exchanged even when not touching sticks and idle.
Yep, everything has to be the same baud and nothing is at 420000. If you scroll up in this thread a couple people including myself had an issue where 420k baud was being set even with airport enabled.

In user_defines.txt uncomment the line that sets baud rate and set it to 9600 or whatever you're using. Then reflash TX/RX. Then enable wifi on both devices and check the options tab to make sure baud is set correct. On the TX you might even see a checkbox saying 'use as airport device' and under it you can set baud.

If you can find me on the elrs discord channel, my username is the same as my github. Send me a message there and I'll see if I can help from there

**Comment by kielerrr** [comment]
> -DUSE_AIRPORT_AT_BAUD=4800

@SkylarTheGrey it looks like a PR was just merged addressing 4800baud issues. https://github.com/ExpressLRS/ExpressLRS/pull/2216

**Comment by teslahed** [comment]
I have now encountered a new problem. I can flash my (betaFPV 900mhz) equipment with expressLRS no issues. I've reflashed it all with the latest versions via the expresslrs configurator program. I have set the baudrate to 4800 in the setup config. I'm trying to set the baudrate to 4800 via the receiver's website gui config page because I've read that's also necessary. The receivers webpage has no tabs and no option to vary the baudrate like I've seen previously. Has this been removed from the airport fork or has something gone wrong for me?
https://postimg.cc/BtqTNGKf

**Comment by pkendall64** [comment]
> I have now encountered a new problem. I can flash my (betaFPV 900mhz) equipment with expressLRS no issues. I've reflashed it all with the latest versions via the expresslrs configurator program. I have set the baudrate to 4800 in the setup config. I'm trying to set the baudrate to 4800 via the receiver's website gui config page because I've read that's also necessary. The receivers webpage has no tabs and no option to vary the baudrate like I've seen previously. Has this been removed from the airport fork or has something gone wrong for me? 

You need to have flashed the master branch, not the PR as that has been merged into master.

**Comment by teslahed** [comment]
So the new procedure is just to flash regular expressLRS and then enable airport with the  -DUSE_AIRPORT_AT_BAUD=whatever comment?

**Comment by pkendall64** [comment]
It's even easier than that now. Just flash it as usual and go to the wifi page to enable airport mode and set the baud rate on the options page.

**Comment by andrzejroszkowski** [comment]
Is there a way to transmit receiver to receiver with this?

**Comment by pkendall64** [comment]
> Is there a way to transmit receiver to receiver with this? Instead of regular TX - link two receivers.

Yes, you can flash TX firmware to an RX and make the required adjustments, like uploading hardware.json file etc.

**Comment by criegit** [comment]
Hi,

i just tried to setup everything for airport use and failed for 2 days now.

Maybe you can have a look if i made a mistake? 

1) i got new TX (BetaFPV micro 1W 2.4ghz) and RX (BetaFPV SuperD 2.4Ghz)
2) i flashed the newest 3.3.0 rc2 firmware on both including EU-CE checked and a bind phrase:
user defines TX:
-DRegulatory_Domain_EU_CE_2400
-DMY_BINDING_PHRASE="XXX"
-DUART_INVERTED
-DAUTO_WIFI_ON_INTERVAL="60"
-DFAST_SYNC

user defines RX:
-DRegulatory_Domain_EU_CE_2400
-DMY_BINDING_PHRASE="XXX"
-DLOCK_ON_FIRST_CONNECTION
-DAUTO_WIFI_ON_INTERVAL="60"
-DRCVR_UART_BAUD="9600"
-DFAST_SYNC

3)I set the TX DIP switches 3&4 on.
4) i set up 333Hz Full 
5) connected to wifi and changed TX & RX to use airport and 9600baud
6) Rebooted everything: it connects immediately. (OLED display on TX)
7) Connected RX to UART. (i do get some readable message when it is going into wifi mode, so i think it is working.)
8) TX is powered only via USB and connected to putty as well.
-> no connection when sending characters on either side. TX is actually rebooting quite often.

i tried:
- different air rate: 500Hz
- UART not inverted on TX 
- tried -DUSE_AIRPORT_AT_BAUD=9600
- tried arduplane with pixhawk and MP -> no connection to plane, but no error connectiong to the com port.
and all combinations.

did i do any obvious mistake?

Plans for afternoon:
- Test alternative power supply on tx,

Thanks!

**Comment by criegit** [comment]
Found my mistake here: https://github.com/ExpressLRS/ExpressLRS/issues/2249

WIth the betaFPV TX you have to keep the dip switches in the Flash configuration: DIP 1&2 ON. Then serial adapter works. Will now go forward with ardupilot...

**Comment by Retlsnake** [comment]
Hello,
I am trying to operate the Airport on TX Jumper 2.4g elrs aion nano and RX Happymodel EP1, communication only works in the direction from TX to RX but not the other way around. I do not know where the problem is. I tried two different RX EP1 and the problem is the same.
I suspect that the TX should be somehow HW modified for the same function, but I'm not sure. I have TX connected to a PC terminal at 9600 baud using USB to TX. I have the RX connected to the laptop via a USB UART converter to the terminal at 9600baud. I have Airport and 9600 booths active in the RX and TX web settings. Communication goes only from XT to RX. Where do I have a problem, please advise. Thank you.

**Comment by akhodakivskiy** [comment]
I'm trying to use Airport with HappyModel ES900 RX/TX with no avail. EX900TX has USB port inside the module (need to unscrew and open). Should this port work as expected while connecting to the ground station? 

Also as I understand the top and bottom JR module pins correspond to the UART TX/RX. Should it be possible to use USB-TTL adapter and wire those pins directly instead of using the USB port?

Is there a short list of Tx/Rx modules that support Airport out of the box without much hassle?

I've tried R9M TX module - it has TX/RX pins on the case (althogh dunno where they go..). Anyone got FrSky stuff to work with Airport?

**Comment by kevinuav** [comment]
Hi,

> ### What is Airport?
> Airport provides a new firmware option that allows you to turn a regular ExpressLRS TX and RX pair into a bi-directional transparent serial link, over the air. The TX module is intended to connect via USB to a laptop, and the RX connects to a free UART on your FC, as per usual. This then allows serial comms between the peers, in any protocol you wish to use, for example: MAVLINK (Ardupilot), MSP (Betaflight and INAV), or any other telemetry protocol you might use.
> 
> ### What hardware do I need to try this?
> The Airport option completely replaces the RC link, and repurposes it as a data link. This means you need to run 2x TXs and 2x RXs on the ground and air, respectively. One TX+RX pair sends your normal RC link data, just as it does it does now, and the other TX+RX pair sends the serial data. airport-diagram
> 
> The feature should compile for any supported ELRS target, however the TX module makes sense to be an external one, so you can connect it via USB to a laptop, and it is helpful to have an OLED screen on the TX module so you can change packet rates / power output without plugging it into a handset JR bay. It should compile for 868/900/2.4 targets, on both STM and ESP architectures, so all RXs with a full UART (both tx and rx pads) should work.
> 
> ### What baud rate should I use?
> The baud rate you select to use should be the same on the TX and RX, and it should also be slow enough such that the packet rate can send the data. For example, at a packet rate of 500hz LoRa, and a 1:2 hard-coded TLM rate, we can send 500hz / 2 * 5bytes = 1250bytes/s. A baud rate of 19200 allows for a data rate of 19200 * 0.8 / 8bits/byte = 1920bytes/s. This is too much data for 500hz, since 1920 > 1250. A baud rate of 9600 allows for a data rate of 9600* 0.8 / 8bits/byte = 960bytes/s. This is an acceptable rate since 960 < 1250.
> 
> ### How do I test this PR?
> Long term, this will be an option you can enable in the ELRS Configurator. For now though, the user define should be used via the manual FW build process (read: dev friendly only ATM). Enable the `-DUSE_AIRPORT_AT_BAUD=9600` option by removing the # at the beginning of the line in `user_defines.txt`, and set your binding phrase user define to something different to what you use for normal ELRS RC control. Build and flash a TX and RX pair using the above options, and connect RX to a spare UART, and TX to a PC via USB. Configure your FC to talk via whatever protocol you prefer, for example if using Ardupilot, configure the UART for MAVLINK protocol. Open the connection from the Configurator / GCS system on the PC. Profit!

Hi, thank you for your work. And I have a suggestion. Why can't we make the TX module sending the telemetry data via wifi in the next version of firmware? This could make the cell phone/pad a ground control station directly. And we don't need the laptop in the fly field any more.

**Pull request merged**